packages/commonwealth/server/routes/bulkBalances.ts

```diff
@@ -0,0 +1,172 @@
+import { AppError, ServerError } from '../util/errors';
+import { success, TypedRequestBody, TypedResponse } from '../types';
+
+enum BulkBalancesErrors {
+  InvalidToken = 'Invalid token',
+  NoProfileId = 'No profileId provided',
+  NoProfile = "Profile doesn't exist",
+  NoRegisteredAddresses = 'No wallet addresses registered to this profile',
+  Failed = 'Request Failed',
+}
+
+// ----------- OUTSTANDING QUESTIONS -----------
+// 1. Max across addresses or sum across addresses?
+// 2. How is this route being authenticated, if at all?
+
+type bulkBalanceReq = {
+  profileId: number;
+  token: string;
+  chainNodes: {
+    [nodeId: number]: string | string[];
+  };
+};
+
+type bulkBalanceResp = {
+  balances: {
+    [nodeId: string]:
+      | {
+          [tokenAddress: string]: number;
+        }
+      | number;
+  };
+  bases: string[];
+};
+
+const bulkBalances = async (
+  models: DB,
+  tokenBalanceCache: TokenBalanceCache,
+  req: TypedRequestBody<bulkBalanceReq>,
+  res: TypedResponse<bulkBalanceResp>
+) => {
+  const { profileId, chainNodes, token } = req.body;
+
+  if (!process.env.DISCORD_BOT_TOKEN)
+    throw new AppError(BulkBalancesErrors.Failed);
+
+  if (!token || token !== process.env.DISCORD_BOT_TOKEN)
+    throw new AppError(BulkBalancesErrors.InvalidToken);
+
+  if (!profileId) throw new AppError(BulkBalancesErrors.NoProfileId);
+
+  const profile = await models.Profile.findOne({
+    where: { id: profileId },
+  });
+  if (!profile) throw new ServerError(BulkBalancesErrors.NoProfile);
+
+  // Get all addresses registered with user
+  const addressInstances: AddressInstance[] = await models.Address.findAll({
+    where: { profile_id: profileId },
+  });
+
+  // Remove duplicates from addresses
+  const profileWalletAddresses: string[] = [
+    ...new Set(addressInstances.map((addressObj) => addressObj.address)),
+  ];
+
+  if (profileWalletAddresses.length === 0) {
+    throw new AppError(BulkBalancesErrors.NoRegisteredAddresses);
+  }
+
+  // get bases for addresses being returned
+  const baseQuery = `
+    SELECT DISTINCT(c.base) FROM "Addresses" addr 
+      LEFT JOIN "Chains" c ON addr.chain = c.id 
+      WHERE addr.profile_id = ${profileId};`;
+  const basesRaw: string[] = <any>(await sequelize.query(baseQuery, {
+    raw: true,
+    type: QueryTypes.SELECT,
+  }));
+
+  const bases = basesRaw.map((b) => b['base']);
+
+  const balances: {
+    [nodeId: string]:
+      | {
+          [tokenAddress: string]: number;
+        }
+      | number;
+  } = {};
+
+  // Iterate through chain nodes
+  for (const nodeIdString of Object.keys(chainNodes)) {
+    let tokenAddresses = req.body.chainNodes[nodeIdString];
+    const nodeId = parseInt(nodeIdString, 10);
+
+    // Handle when only one value in array
+    if (typeof tokenAddresses === 'string') {
+      tokenAddresses = [tokenAddresses];
+    }
+
+    // Handle no token addresses for chain node
+    if (!tokenAddresses || tokenAddresses.length === 0) {
```

Comment:
Lot of repeated code in here -- only differences are how we're invoking TBC.getBalance

Reply:
what do you think we should do?

Reply:
since we're trying to get it out the door, it's not a big deal if we leave it like this. just noting that it's not amazing style.